File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaqueDistribuicaoPorPosicao.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import fbc.Seguranca.SkipAtaqueException;
import fbc.Seguranca.Util;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/**
 * O AtaqueDistribuicaoBytes faz qui-quadrado GLOBAL. Esse faz POR POSIÇÃO do
 * bloco de 32 bytes: uma posição específica pode ter viés forte mesmo que a
 * soma global pareça uniforme (o viés de uma posição se dilui entre as 32).
 */
public class AtaqueDistribuicaoPorPosicao implements AtaqueInterface {
    private final int amostrasDeBlocos;
    private final int tamanhoBloco;

    public AtaqueDistribuicaoPorPosicao() {
        this(2000, 32);
    }

    public AtaqueDistribuicaoPorPosicao(int amostrasDeBlocos, int tamanhoBloco) {
        this.amostrasDeBlocos = amostrasDeBlocos;
        this.tamanhoBloco = tamanhoBloco;
    }

    @Override
    public String nome() {
        return "Distribuição de bytes por posição do bloco";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        byte[] chave = alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8);
        byte[] primeiro = alvo.gerarKeystreamBruto(
                chave, Util.randomBytes(alvo.tamanhoIv()), "enc", this.tamanhoBloco);
        if (primeiro == null) {
            throw new SkipAtaqueException("Alvo não expõe gerarKeystreamBruto.");
        }

        int[][] contagens = new int[this.tamanhoBloco][256];
        for (int i = 0; i < this.amostrasDeBlocos; i++) {
            byte[] ks = alvo.gerarKeystreamBruto(
                    chave, Util.randomBytes(alvo.tamanhoIv()), "enc", this.tamanhoBloco);
            for (int p = 0; p < this.tamanhoBloco; p++) {
                contagens[p][ks[p] & 0xFF]++;
            }
        }

        double esperado = (double) this.amostrasDeBlocos / 256;
        List<String> problemas = new ArrayList<>();
        double pior = 0.0;

        for (int p = 0; p < contagens.length; p++) {
            int[] c = contagens[p];
            double chi2 = 0.0;
            for (int v : c) {
                double d = v - esperado;
                chi2 += (d * d) / esperado;
            }
            pior = Math.max(pior, chi2);
            double z = (chi2 - 255) / Math.sqrt(510);
            if (z > 4.0) {
                problemas.add(String.format(Locale.ROOT, "posição %d chi2=%.1f", p, chi2));
            }
        }

        if (!problemas.isEmpty()) {
            return new ResultadoAtaque(nome(), true, Severidade.MEDIA, String.join("; ", problemas));
        }

        return new ResultadoAtaque(
                nome(),
                false,
                Severidade.INFO,
                String.format(
                        Locale.ROOT,
                        "Todas as %d posições uniformes (pior chi2=%.1f, esperado ~255)",
                        this.tamanhoBloco,
                        pior));
    }
}

File: packages/go/internal/seguranca/ataques/ataque_distribuicao_por_posicao.go
package ataques

import (
	"fmt"
	"math"
	"strings"

	"criptografia/internal/seguranca"
)

// OpcoesDistribuicaoPorPosicao configura o AtaqueDistribuicaoPorPosicao.
type OpcoesDistribuicaoPorPosicao struct {
	AmostrasDeBlocos int
	TamanhoBloco     int
}

// AtaqueDistribuicaoPorPosicao faz qui-quadrado POR POSIÇÃO do bloco, onde
// uma posição específica pode ter viés forte mesmo que a soma global pareça
// uniforme.
type AtaqueDistribuicaoPorPosicao struct {
	amostrasDeBlocos int
	tamanhoBloco     int
}

// NovoAtaqueDistribuicaoPorPosicao cria o ataque. Sem argumento, usa
// amostrasDeBlocos = 2000 e tamanhoBloco = 32.
func NovoAtaqueDistribuicaoPorPosicao(opcoes ...OpcoesDistribuicaoPorPosicao) *AtaqueDistribuicaoPorPosicao {
	o := OpcoesDistribuicaoPorPosicao{AmostrasDeBlocos: 2000, TamanhoBloco: 32}
	if len(opcoes) > 0 {
		o = opcoes[0]
	}
	return &AtaqueDistribuicaoPorPosicao{amostrasDeBlocos: o.AmostrasDeBlocos, tamanhoBloco: o.TamanhoBloco}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueDistribuicaoPorPosicao) Nome() string {
	return "Distribuição de bytes por posição do bloco"
}

// Executar roda o ataque.
func (a *AtaqueDistribuicaoPorPosicao) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	chave := []byte(alvo.ChaveDeTeste())

	contagens := make([][]int, a.tamanhoBloco)
	for p := 0; p < a.tamanhoBloco; p++ {
		contagens[p] = make([]int, 256)
	}

	for i := 0; i < a.amostrasDeBlocos; i++ {
		ks := alvo.GerarKeystreamBruto(chave, seguranca.RandomBytes(alvo.TamanhoIv()), "enc", a.tamanhoBloco)
		for p := 0; p < a.tamanhoBloco; p++ {
			contagens[p][ks[p]]++
		}
	}

	esperado := float64(a.amostrasDeBlocos) / 256
	problemas := []string{}
	pior := 0.0

	for p := 0; p < len(contagens); p++ {
		chi2 := 0.0
		for _, v := range contagens[p] {
			diff := float64(v) - esperado
			chi2 += (diff * diff) / esperado
		}
		if chi2 > pior {
			pior = chi2
		}
		z := (chi2 - 255) / math.Sqrt(510)
		if z > 4.0 {
			problemas = append(problemas, fmt.Sprintf("posição %d chi2=%.1f", p, chi2))
		}
	}

	if len(problemas) > 0 {
		return seguranca.ResultadoAtaque{
			NomeAtaque: a.Nome(),
			Vulneravel: true,
			Severidade: seguranca.SevMedia,
			Detalhes:   strings.Join(problemas, "; "),
		}, nil
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: false,
		Severidade: seguranca.SevInfo,
		Detalhes: fmt.Sprintf(
			"Todas as %d posições uniformes (pior chi2=%.1f, esperado ~255)",
			a.tamanhoBloco, pior,
		),
	}, nil
}
